File: chapter_2/main_body/eight_queen.c
#include <stdio.h>
int Queens[9]={0},Count=0;
int isplaceok(int row,int list);
void printsolution(void);
void addqueen(int row);
int main(void)
{
	addqueen(1);
	printf("Count=%d\n",Count);
	return 0;
}
int isplaceok(int row,int list)
{
	for(int index=1;index<row;index++)
	{
		if((Queens[index]==list)||(Queens[index]-index==list-row)||(Queens[index]+index==row+list))
			return 0;
	}
	return 1;
}
void printsolution(void)
{
	for(int i=1;i<9;i++)
	{
		for(int j=1;j<9;j++)
		{
			if(Queens[i]==j)
				printf("X");
			else
				printf("-");
		}
		putchar('\n');
	}
	putchar('\n');
}
void addqueen(int row)
{
	for(int list=1;list<9;list++)
	{
		if(isplaceok(row,list))
		{
			Queens[row]=list;
			if(row==8)
			{
				printsolution();
				Count++;
				return;
			}
			addqueen(row+1);
		}
	}
}

File: chapter_2/main_body/eight_queen.lua
#!/usr/bin/lua
N=8
--检测n,c是否不会被以存在的皇后攻击
function isplaceok(a,n,c)
	for i=1,n-1 do
		if (a[i]==c) or (a[i]-i==c-n) or (a[i]+i==c+n) then
			return false
		end
	end
	return true
end
--打印棋盘
function printsolution(a)
	for i=1,N do
		for j=1,N do
			io.write(a[i]==j and "X" or "-","")
		end
		io.write("\n")
	end
	io.write("\n")
end
--把从第n行开始一直到N行的所有皇后放入棋盘
function addqueen(a,n)
	if (n>N) then
		printsolution(a)
	else
		for c=1,N do
			if(isplaceok(a,n,c)) then
				a[n]=c
				addqueen(a,n+1)
			end
		end
	end
end
addqueen({},1)
